model/common/src/icon4py/model/common/metrics/metrics_factory.py

```diff
@@ -0,0 +1,554 @@
+edge_domain = horizontal.domain(dims.EdgeDim)
+vertex_domain = horizontal.domain(dims.VertexDim)
+#######
+
+# start build up factory:
+
+# used for vertical domain below: should go away once vertical grid provids start_index and end_index like interface
+grid = grid_savepoint.global_grid_params
+
+interface_model_height = metrics_savepoint.z_ifc()
+c_lin_e = interpolation_savepoint.c_lin_e()
+k_index = gtx.as_field((dims.KDim,), xp.arange(nlev + 1, dtype=gtx.int32))
+vct_a = grid_savepoint.vct_a()
+theta_ref_mc = metrics_savepoint.theta_ref_mc()
+exner_ref_mc = metrics_savepoint.exner_ref_mc()
+wgtfac_c = metrics_savepoint.wgtfac_c()
+c_refin_ctrl = grid_savepoint.refin_ctrl(dims.CellDim)
+e_refin_ctrl = grid_savepoint.refin_ctrl(dims.EdgeDim)
+dual_edge_length = grid_savepoint.dual_edge_length()
+tangent_orientation = grid_savepoint.tangent_orientation()
+inv_primal_edge_length = grid_savepoint.inverse_primal_edge_lengths()
+inv_dual_edge_length = grid_savepoint.inv_dual_edge_length()
+cells_aw_verts = interpolation_savepoint.c_intp().asnumpy()
+cells_aw_verts_field = gtx.as_field((dims.VertexDim, dims.V2CDim), cells_aw_verts)
+vwind_offctr = 0.2
+icon_grid = grid_savepoint.construct_icon_grid(on_gpu=False)
+vwind_impl_wgt_full = constant_field(icon_grid, 0.5 + vwind_offctr, dims.CellDim)
+experiment = dt_utils.GLOBAL_EXPERIMENT
+init_val = 0.65 if experiment == dt_utils.GLOBAL_EXPERIMENT else 0.7
+vwind_impl_wgt_k = constant_field(icon_grid, init_val, dims.CellDim, dims.KDim)
+k_lev = gtx.as_field((dims.KDim,), np.arange(nlev, dtype=gtx.int32))
+e_lev = gtx.as_field((dims.EdgeDim,), np.arange(icon_grid.num_edges, dtype=gtx.int32))
+e_owner_mask = grid_savepoint.e_owner_mask()
+
+fields_factory = factory.FieldsFactory()
+
+fields_factory.register_provider(
+    factory.PrecomputedFieldsProvider(
+        {
+            "height_on_interface_levels": interface_model_height,
+            "cell_to_edge_interpolation_coefficient": c_lin_e,
+            cf_utils.INTERFACE_LEVEL_STANDARD_NAME: k_index,
+            "vct_a": vct_a,
+            "theta_ref_mc": theta_ref_mc,
+            "exner_ref_mc": exner_ref_mc,
+            "wgtfac_c": wgtfac_c,
+            "c_refin_ctrl": c_refin_ctrl,
+            "e_refin_ctrl": e_refin_ctrl,
+            "dual_edge_length": dual_edge_length,
+            "tangent_orientation": tangent_orientation,
+            "inv_primal_edge_length": inv_primal_edge_length,
+            "inv_dual_edge_length": inv_dual_edge_length,
+            "cells_aw_verts_field": cells_aw_verts_field,
+            "vwind_impl_wgt_full": vwind_impl_wgt_full,
+            "vwind_impl_wgt_k": vwind_impl_wgt_k,
+            "k_lev": k_lev,
+            "e_lev": e_lev,
+            "e_owner_mask": e_owner_mask,
+        }
+    )
+)
+
+
+height_provider = factory.ProgramFieldProvider(
+    func=mf.compute_z_mc,
+    domain={
+        dims.CellDim: (
+            horizontal._local(dims.CellDim),
+            horizontal._end(dims.CellDim),
```

Comment:
that should become

```
dims.CellDim: (h_grid.domain(dims.CellDim)(h_grid.Zone.LOCAL), h_grid.domain(dims.CellDim)(h_grid.Zone.END)),
```
if you have to use a `_func` from another module then you should always be suspicious because they are private... But for the above to work, the factory itself needs to be adapted. I pushed that to the prototype branch.

Reply:
That is why we also need a better solution for the params that is the scalar args, as most of them are also not known at declaration time. That is what I meant with the "more such abstractions"